File: app/src/main/java/de/project/lukas/model/RemoteController.java
package de.project.lukas.model;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import java.util.Locale;

public interface RemoteController {
    void up(Remote remote);

    void down(Remote remote);

    void middle(Remote remote);

    LiveData<String> getName();

    static RemoteController noop() {
        return NoopController.getInstance();
    }

    class MotorController implements RemoteController {
        private final TrainHub device;
        private final LiveData<String> name;

        @Override
        public LiveData<String> getName() {
            return name;
        }

        MotorController(TrainHub device) {
            this.device = device;

            name = Transformations.map(device.getName(),
                    n -> String.format(Locale.getDefault(), "Motor %s", n));
        }

        public void up(Remote remote) {
            device.motorFaster();
        }

        public void down(Remote remote) {
            device.motorSlower();
        }

        public void middle(Remote remote) {
            device.motorStop();
        }
    }

    class BaseMotorController implements RemoteController {
        private final TrainBase device;
        private final LiveData<String> name;

        @Override
        public LiveData<String> getName() {
            return name;
        }

        BaseMotorController(TrainBase device) {
            this.device = device;

            name = Transformations.map(device.getName(),
                    n -> String.format(Locale.getDefault(), "Motor %s", n));
        }

        public void up(Remote remote) {
            device.motorFaster();
        }

        public void down(Remote remote) {
            device.motorSlower();
        }

        public void middle(Remote remote) {
            device.motorStop();
        }
    }

    class LightController implements RemoteController {
        private final TrainHub device;
        private final LiveData<String> name;

        @Override
        public LiveData<String> getName() {
            return name;
        }

        LightController(TrainHub device) {
            this.device = device;

            name = Transformations.map(device.getName(),
                    n -> String.format(Locale.getDefault(), "Light %s", n));
        }

        public void up(Remote remote) {
            device.lightBrighter();
        }

        public void down(Remote remote) {
            device.setLedColorHub();
            remote.setLedColorRemote(device.getCurrentColor());
        }

        public void middle(Remote remote) {
            device.lightOff();
        }
    }

    class BaseLightController implements RemoteController {
        private final TrainBase device;
        private final LiveData<String> name;

        @Override
        public LiveData<String> getName() {
            return name;
        }

        BaseLightController(TrainBase device) {
            this.device = device;

            name = Transformations.map(device.getName(),
                    n -> String.format(Locale.getDefault(), "Light %s", n));
        }

        public void up(Remote remote) {
            // device.lightBrighter();
        }

        public void down(Remote remote) {
            // device.lightDarker();
        }

        public void middle(Remote remote) {
            device.setLedColorHub();
        }
    }

    class SwitchController implements RemoteController {
        private final Switch device;
        private final LiveData<String> name;

        @Override
        public LiveData<String> getName() {
            return name;
        }

        SwitchController(Switch device) {
            this.device = device;

            name = Transformations.map(device.getName(),
                    n -> String.format(Locale.getDefault(), "Switch %s", n));
        }

        public void up(Remote remote) {
            device.toggle1();
        }

        public void middle(Remote remote) {
        }
        public void down(Remote remote) { device.toggle2();
        }
    }

    class NoopController implements RemoteController {
        private static final NoopController INSTANCE = new NoopController();
        private final MutableLiveData<String> name = new MutableLiveData<>();

        public static NoopController getInstance() {
            return INSTANCE;
        }

        @Override
        public LiveData<String> getName() {
            return name;
        }

        private NoopController() {
            name.setValue("None");
        }

        public void up(Remote remote) {
        }

        public void down(Remote remote) {
        }

        public void middle(Remote remote) {
        }
    }
}

File: app/src/main/java/de/project/lukas/model/RemoteController.kt
package de.project.lukas.model

/**
 * Strategy that maps a physical remote's three buttons (up / down / middle) onto an action on a
 * target device. This is what makes the app a mediator: a remote can drive any connected device.
 */
interface RemoteController {
    fun up(remote: Remote)
    fun down(remote: Remote)
    fun middle(remote: Remote)

    /** Snapshot label for the selection dropdowns. */
    val displayName: String

    companion object {
        val Noop: RemoteController = NoopController
    }
}

/** Does nothing; used when a remote button is unbound. */
object NoopController : RemoteController {
    override fun up(remote: Remote) {}
    override fun down(remote: Remote) {}
    override fun middle(remote: Remote) {}
    override val displayName: String get() = "None"
}

class MotorController(private val device: TrainHub) : RemoteController {
    override fun up(remote: Remote) = device.motorFaster()
    override fun down(remote: Remote) = device.motorSlower()
    override fun middle(remote: Remote) = device.motorStop()
    override val displayName: String get() = "Motor ${device.name.value}"
}

class LightController(private val device: TrainHub) : RemoteController {
    override fun up(remote: Remote) = device.lightBrighter()
    override fun down(remote: Remote) {
        device.setLedColorHub()
        remote.setLedColorRemote(device.currentColor)
    }

    override fun middle(remote: Remote) = device.lightOff()
    override val displayName: String get() = "Light ${device.name.value}"
}

class BaseMotorController(private val device: TrainBase) : RemoteController {
    override fun up(remote: Remote) = device.motorFaster()
    override fun down(remote: Remote) = device.motorSlower()
    override fun middle(remote: Remote) = device.motorStop()
    override val displayName: String get() = "Motor ${device.name.value}"
}

class BaseLightController(private val device: TrainBase) : RemoteController {
    override fun up(remote: Remote) {}
    override fun down(remote: Remote) {}
    override fun middle(remote: Remote) = device.setLedColorHub()
    override val displayName: String get() = "Light ${device.name.value}"
}

class SwitchController(private val device: Switch) : RemoteController {
    override fun up(remote: Remote) = device.toggle1()
    override fun down(remote: Remote) = device.toggle2()
    override fun middle(remote: Remote) {}
    override val displayName: String get() = "Switch ${device.name.value}"
}
